Sources/SourceControl/GitRepository.swift

```diff
@@ -33,9 +33,59 @@ public class GitRepositoryProvider: RepositoryProvider {
+    private let cachePath: AbsolutePath
+    /// The maximum size of the cache in bytes.
+    private let maxCacheSize: UInt64
+
+    public init(processSet: ProcessSet? = nil,
+                cachePath: AbsolutePath = AbsolutePath(localFileSystem.homeDirectory, ".swiftpm/cache/repositories"),
+                maxCacheSize: UInt64 = 21_474_836_480) {
         self.processSet = processSet
+        self.cachePath = cachePath
+        self.maxCacheSize = maxCacheSize
+    }
+
+    /// Clones the  git repository we want to cache into the cache directory if it does not already exist and returns it.
+    private func setupCacheIfNeeded(for repository: RepositorySpecifier) throws -> GitRepository {
+        let repositoryPath = cachePath.appending(component: repository.fileSystemIdentifier)
+
+        guard !localFileSystem.exists(repositoryPath) else { return GitRepository(path: repositoryPath, isWorkingRepo: false) }
+
+        try localFileSystem.createDirectory(repositoryPath, recursive: true)
+
+        // We are cloning each repository into its own directory instead of using one large bare repository and adding a remote for each repository.
+        // This avoids the large overhead that occurs when git tries to determine if it has any revision in common with the remote repository,
+        // which involves sending a list of all local commits to the server (a potentially huge list depending on cache size
+        // with most commits unrelated to the repository we actually want to fetch).
+        try Process.checkNonZeroExit(
+            args: Git.tool, "clone", "--mirror", repository.url, repositoryPath.pathString)
+
+        return GitRepository(path: repositoryPath, isWorkingRepo: false)
+    }
+
+    /// Purges git repositories from the cache directory in order to free some space.
+    private func purgeCacheIfNeeded() {
+        do {
+            let cacheSize = try localFileSystem.getDirectorySize(cachePath)
+            let desiredCacheSize = maxCacheSize - (maxCacheSize / 8)
+
+            guard cacheSize > maxCacheSize else { return }
+
+            let repositories = try localFileSystem.getDirectoryContents(cachePath)
+                .map { GitRepository(path:  cachePath.appending(component: $0), isWorkingRepo: false) }
+                .sorted { (try localFileSystem.getFileInfo($0.path).modTime) < (try localFileSystem.getFileInfo($1.path).modTime) }
+
+            // Purges repositories until the desired cache size is reached.
+            for repository in repositories {
+                let cacheSize = try localFileSystem.getDirectorySize(cachePath)
+                guard cacheSize > desiredCacheSize else { break }
+                try localFileSystem.removeFileTree(repository.path)
+            }
+        } catch {
+            // The cache seems to be broken. Lets remove everything.
+            try? localFileSystem.removeFileTree(cachePath)
+        }
```

Comment:
I think the error should still be reported somehow, maybe just printing it would be an improvement.